src/main/java/org/jenkinsci/plugins/saml/SamlSecurityRealm.java

```diff
@@ -374,22 +425,27 @@ private String getUsernameFromProfile(Saml2Profile saml2Profile) {
    * they can configure their IdP.
    */
   public HttpResponse doMetadata(StaplerRequest request, StaplerResponse response) {
-    Saml2Client client = newClient();
+    final SAML2Client client = newClient(request,response);
 
-    return HttpResponses.plainText(client.printClientMetadata());
+    return HttpResponses.plainText(client.getServiceProviderMetadataResolver().getMetadata());
   }
 
-  private Saml2Client newClient() {
+  private SAML2Client newClient(StaplerRequest request, StaplerResponse response) {
     Preconditions.checkNotNull(idpMetadata);
+    final SAML2ClientConfiguration client = new SAML2ClientConfiguration();
+    client.setIdentityProviderMetadataResource(new RewriteableStringResource(idpMetadata));
+    client.setServiceProviderMetadataResource(new RewriteableStringResource());
 
-    Saml2Client client = new Saml2Client();
-    client.setIdpMetadata(idpMetadata);
-    client.setCallbackUrl(getConsumerServiceUrl());
-    client.setDestinationBindingType(SAMLConstants.SAML2_REDIRECT_BINDING_URI);
```

Comment:
maybe it is better to add a field to define the binding method, to allow the user to decided.